models/deeplab/evaluate.py

```diff
@@ -115,14 +115,38 @@ def test(self, ):
                 total_recall.append(0)
 
         print({
-                "test_mIOU": np.mean(total_mIOU),
-                "test_pixel_acc": np.mean(total_pixelAcc),
-                "test_f1": np.mean(total_f1),
-                "test_ap": np.mean(total_precision),
-                "test_ar": np.mean(total_recall),
-            })
+            "test_mIOU": np.mean(total_mIOU),
+            "test_pixel_acc": np.mean(total_pixelAcc),
+            "test_f1": np.mean(total_f1),
+            "test_ap": np.mean(total_precision),
+            "test_ar": np.mean(total_recall),
+        })
+
+    def inference(self, input_width, input_height):
+        assert self.test_loader.dataset.__class__.__name__ == "TileInferenceDataset"
+
+        tbar = tqdm(self.test_loader)
+
+        output = np.zeros((2, input_height, input_width), dtype=np.float32)
+        kernel = np.ones((self.args.chip_size, self.args.chip_size), dtype=np.float32)
+        kernel[self.args.padding:-self.args.padding, self.args.padding:-self.args.padding] = 5
+        counts = np.zeros((input_height, input_width), dtype=np.float32)
 
+        for i, (data, coords) in enumerate(tbar):
+            if self.args.cuda:
+                data = data.cuda()
+
+            with torch.no_grad():
+                t_output = self.model(data)
+                t_output = torch.nn.functional.softmax(t_output, dim=1).cpu().numpy()
 
+            for j in range(t_output.shape[0]):
+                y, x =  coords[j]
 
+                output[:, y:y+self.args.chip_size, x:x+self.args.chip_size] += t_output[j] * kernel
+                counts[y:y+self.args.chip_size, x:x+self.args.chip_size] += kernel
 
+        output = output / counts
```

Comment:
What is `counts` used for? Is this if a particular set of pixels were used multiple times for inference, it would average out the predicted classes?

Reply:
Yep exactly -- because of the stride some pixels can be predicted in multiple patches. This doesn't matter if you just want the argmax class, but does matter if you want to save the probabilities.